### ISSUE TITLE
Let `tox --listenvs` list all defined environments, not only from `envlist` in `[tox]` section

### ISSUE THREAD

**Issue by pytoxbot** [tox-dev/tox#271]
- Bitbucket: https://bitbucket.org/hpk42/tox/issue/271
- Originally reported by: [vlcinsky](http://bitbucket.org/vlcinsky)
- Originally created at: 2015-08-28T00:01:52.447

Issue #52 asks for having a command to list all configured environments.

However, I found, that it lists only those environments, which are in `envlist`.

If I do not want to run some of the environments when `tox` is invoked without any argument, I keep the `envlist` in `[tox]` shorter. However, this does not print it out when::

```
$ tox -l
```

It would be nice to list all existing environments.

tox version: 2.1.1

OS: Debian Jessie

Python 2.7.9

Sample `tox.ini`:

```
#!ini

# Tox (http://tox.testrun.org/) is a tool for running tests
# in multiple virtualenvs. This configuration file will run the
# test suite on all supported python versions. To use it, "pip install tox"
# and then run "tox" from this directory.

[tox]
envlist = help
skipsdist = true

[testenv:help]
basepython = python2.7
commands = python -c "print(open('tox.help').read());" 
deps =

[testenv:docs]
basepython = python2.7
whitelist_externals =
    make
changedir = docs
commands = make html
deps = 
    -rrequirements-docs.txt

[testenv:testsamples]
basepython = python2.7
envdir = {toxworkdir}/test
commands = py.test -sv tests/test_samples.py
deps =
    -rrequirements-json.txt

[testenv:testschema]
basepython = python2.7
envdir = {toxworkdir}/test
commands = py.test -sv tests/test_schema.py
deps =
    -rrequirements-json.txt

[testenv:testyaml]
basepython = python2.7
envdir = {toxworkdir}/test
commands = py.test -sv tests/test_sample.py --yamlsample []
deps =
    -rrequirements-json.txt

[testenv:testjson]
basepython = python2.7
envdir = {toxworkdir}/test
commands = py.test -sv tests/test_sample.py --jsonsample []
deps =
    -rrequirements-json.txt

```

For given `tox.ini` it lists only `help`:

```
#!bash

$ tox --listenvs
help
```

and I would like to see there all:

```
#!bash

$ tox --listenvs
help
docs
testsamples
testschema
testyaml
testjson
```

**Comment by volans-** [comment]
+1, how a user could _discover_ an environment not run by default?

**Comment by obestwalter** [comment]
>  how a user could discover an environment not run by default?

Until this issue is resolved only by looking for all `[testenv:???]`sections directly in the `tox.ini` of the project .

**Comment by gaborbernat** [comment]
created an implementation for this based on top of https://github.com/tox-dev/tox/pull/489, I'll create the PR once that gets accepted :+1:

**Comment by obestwalter** [comment]
fixed in #495
